cpp/src/bootstrap/file_backend.cpp

```diff
@@ -0,0 +1,296 @@
+
+    // NFS visibility aid: derive parent directory to refresh its metadata
+    std::string parent_dir;
+    {
+        auto pos = path.find_last_of('/');
+        if (pos != std::string::npos && pos > 0) {
+            parent_dir = path.substr(0, pos);
+        }
+    }
+    auto last_dir_scan = start - std::chrono::milliseconds{1000};
+
+    while (true) {
+        std::error_code ec;
+        if (std::filesystem::exists(path, ec)) {
+            return true;  // File exists
+        }
+
+        auto elapsed = std::chrono::steady_clock::now() - start;
+        if (elapsed >= timeout) {
+            return false;  // Timeout
+        }
+
+        // Hint NFS to refresh directory cache: status and occasionally iterate directory
+        // on parent. Without this remote processes may timeout to spawn because NFS never
+        // refreshes.
+        if (!parent_dir.empty()) {
+            std::ignore = std::filesystem::status(parent_dir, ec);
+
+            auto now = std::chrono::steady_clock::now();
+            if (now - last_dir_scan >= std::chrono::milliseconds{500}) {
+                std::filesystem::directory_iterator it(
+                    parent_dir,
+                    std::filesystem::directory_options::skip_permission_denied,
+                    ec
+                );
+                if (!ec) {
+                    for (; it != std::filesystem::directory_iterator(); ++it) {
+                        std::ignore =
+                            it->path();  // no-op; traversal nudges directory cache
+                    }
+                }
+                last_dir_scan = now;
+            }
+        }
+
+        // Sleep before next poll
+        std::this_thread::sleep_for(poll_interval);
+
+        // Exponential backoff up to 100ms
+        if (poll_interval < std::chrono::milliseconds{100}) {
+            poll_interval = std::min(poll_interval * 2, std::chrono::milliseconds{100});
+        }
+    }
+}
+
+void FileBackend::write_file(std::string const& path, std::string const& content) {
+    std::string tmp_path = path + ".tmp." + std::to_string(getpid());
+
+    // Write to temporary file
+    std::ofstream ofs(tmp_path, std::ios::binary | std::ios::trunc);
+    if (!ofs) {
+        throw std::runtime_error("Failed to open temporary file: " + tmp_path);
+    }
+    ofs << content;
+    ofs.close();
+
+    // Atomic rename
+    {
+        std::error_code ec;
+        std::filesystem::rename(tmp_path, path, ec);
+        if (ec) {
+            std::error_code rm_ec;
+            std::filesystem::remove(tmp_path, rm_ec);  // Clean up temp file
+            throw std::runtime_error(
+                "Failed to rename " + tmp_path + " to " + path + ": " + ec.message()
+            );
+        }
+    }
+}
+
+std::string FileBackend::read_file(std::string const& path) {
+    std::ifstream ifs(path, std::ios::binary);
+    if (!ifs) {
+        throw std::runtime_error("Failed to open file for reading: " + path);
+    }
```

Comment:
Use `RAPIDSMPF_EXPECTS()`

Reply:
Also fixed in https://github.com/rapidsai/rapidsmpf/pull/616/commits/db112cb0245b52cfecc79b956cca86077a20dc47